index/segment/mem/postings_list.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"errors"
+	"math"
+	"sync"
+
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/uber-go/atomic"
+)
+
+var (
+	// ErrOutOfRange is returned when the provided offset does not exist.
+	ErrOutOfRange = errors.New("provided offset does not exist")
+)
+
+type pList struct {
+	sync.RWMutex
+	// TODO(prateek): test perf impact of slice v map here
+	postings []docIDsSet
+
+	counter *atomic.Int32
+}
+
+func newPostingsList(initialSize int) postingsList {
+	return &pList{
+		postings: make([]docIDsSet, initialSize),
+		counter:  atomic.NewInt32(-1),
+	}
+}
+
+func (p *pList) Insert(i segment.DocID) (postingOffset, error) {
+	t := postingOffset(p.counter.Inc())
+	p.ensureSufficientCapacity(t)
+	err := p.Update(t, i)
+	return t, err
+}
+
+func (p *pList) ensureSufficientCapacity(t postingOffset) {
+	p.RLock()
+	size := len(p.postings)
+	p.RUnlock()
+
+	// early terminate if we have sufficient capacity
+	if int(t) < size {
+		return
+	}
+
+	// expand capacity
+	p.Lock()
+
+	// early terminate if we've expanded capacity since releasing lock last
+	if size = len(p.postings); int(t) < size {
+		p.Unlock()
+		return
+	}
+
+	// amortized O(1) expansion
+	factor := math.Ceil(float64(t) / float64(len(p.postings)))
+	if factor < 2 {
```

Comment:
Is it worth expanding the size of the slice linearly past some limit? For example, we always double up to X MB, at which point we expanding capacity X bytes at a time?

Reply:
meh, lets see when we look at allocation profiles